### ISSUE TITLE
CoT Shap values include the answer

### ISSUE THREAD

**Issue by wj210** [Heidelberg-NLP/CC-SHAP#3]
Hi, appreciate you releasing this code, i have some questions regarding the implementation.

For `cc-shap-cot`:

the code averages across the entire sequence length in https://github.com/Heidelberg-NLP/CC-SHAP/blob/2df57423e14a98037e962f84f3152e4fab5f4bf8/faithfulness.py#L174 , including the answer since Cot provides the answer at the end, but should we just compute the average across the explanation instead? Since we are measuring the divg against the output answer, wouldn't this result in some inaccuracies in the metric?

example: 
`['', '', 'S', 'ent', 'ence', '(', 'A', '):', '"', 'She', 'put', 'the', 'papers', 'into', 'the', 'fil', 'ing', 'cabinet', '."', '', '', 'This', 'sentence', 'makes', 'sense', ',', 'as', 'it', 'implies', 'that', 'the', 'subject', '(', 'She', ')', 'placed', 'the', 'objects', '(', 'p', 'apers', ')', 'into', 'a', 'container', '(', 'fil', 'ing', 'cabinet', ').', 'It', "'", 's', 'a', 'common', 'action', 'and', 'a', 'logical', 'sequence', 'of', 'events', '.', '', '', 'S', 'ent', 'ence', '(', 'B', '):', '"', 'She', 'put', 'the', 'fil', 'ing', 'cabinet', 'into', 'the', 'papers', '."', '', '', 'This', 'sentence', 'does', 'not', 'make', 'sense', ',', 'as', 'it', 'implies', 'that', 'the', 'subject', '(', 'She', ')', 'placed', 'a', 'container', '(', 'fil', 'ing', 'cabinet', ')', 'into', 'the', 'objects', '(', 'p', 'apers', ')', 'instead', 'of', 'the', 'other', 'way', 'around', '.', 'It', 'goes', 'against', 'common', 'sense', ',', 'as', 'we', 'wouldn', "'", 't', 'put', 'a', 'large', 'object', 'like', 'a', 'fil', 'ing', 'cabinet', 'into', 'a', 'small', 'p', 'ile', 'of', 'papers', '.', '', '', 'There', 'fore', ',', 'the', 'best', 'answer', 'is', ':', '(', 'A', ')', '"', 'She', 'put', 'the', 'papers', 'into', 'the', 'fil', 'ing', 'cabinet', '."']`

Could we just average until "The best answer..."?

**Comment by LetiP** [comment]
Hello, yes, indeed, sometimes the model generates the answer even in CoT. Sometimes in the form you just showed, but sometimes also by pure paraphrasing. Shortening the generation length can alleviate this, but it also delivers less spaces for the model to reason.
But sure, if you are able to detect these answers, you can cut off the interpretation before them. And the answer setting you just described, is easily detectable.

**Comment by wj210** [comment]
Thanks! i guess thats the purpose of CoT - to generate the answer after conditioning on the explanation ( ideally makes it more faithful).

Also I did a small experiment to see the effects of varying the explanation length (post-hoc), and it seems that cc-shap is pretty robust to the length. I compared between the normal generation and explicitly prompting it to be concise and short. (-0.05 and -0.0332) - not much diff. It seems that the metric is robust against explanations which may be too wordy (ie unimportant infos that might change the distribution of the feature scores). Just letting you know in the case you did not experiment with that :) 

Good work!
